<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API &mdash; FinaleTools  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Releases" href="releases.html" />
    <link rel="prev" title="CLI" href="cli.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            FinaleTools
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quick_start.html">Quick-start Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">CLI</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basic-features">Basic Features</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.coverage"><code class="docutils literal notranslate"><span class="pre">coverage()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.frag_length"><code class="docutils literal notranslate"><span class="pre">frag_length()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.frag_length_bins"><code class="docutils literal notranslate"><span class="pre">frag_length_bins()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#window-protection-score">Window Protection Score</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.wps"><code class="docutils literal notranslate"><span class="pre">wps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.multi_wps"><code class="docutils literal notranslate"><span class="pre">multi_wps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.adjust_wps"><code class="docutils literal notranslate"><span class="pre">adjust_wps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.agg_wps"><code class="docutils literal notranslate"><span class="pre">agg_wps()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#delfi">DELFI</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.delfi"><code class="docutils literal notranslate"><span class="pre">delfi()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.delfi_gc_correct"><code class="docutils literal notranslate"><span class="pre">delfi_gc_correct()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.delfi_merge_bins"><code class="docutils literal notranslate"><span class="pre">delfi_merge_bins()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#end-motifs">End-motifs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.EndMotifFreqs"><code class="docutils literal notranslate"><span class="pre">EndMotifFreqs</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifFreqs.from_file"><code class="docutils literal notranslate"><span class="pre">EndMotifFreqs.from_file()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifFreqs.motif_diversity_score"><code class="docutils literal notranslate"><span class="pre">EndMotifFreqs.motif_diversity_score()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifFreqs.to_tsv"><code class="docutils literal notranslate"><span class="pre">EndMotifFreqs.to_tsv()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.EndMotifsIntervals"><code class="docutils literal notranslate"><span class="pre">EndMotifsIntervals</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifsIntervals.freq"><code class="docutils literal notranslate"><span class="pre">EndMotifsIntervals.freq()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifsIntervals.from_file"><code class="docutils literal notranslate"><span class="pre">EndMotifsIntervals.from_file()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifsIntervals.mds_bed"><code class="docutils literal notranslate"><span class="pre">EndMotifsIntervals.mds_bed()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifsIntervals.motif_diversity_score"><code class="docutils literal notranslate"><span class="pre">EndMotifsIntervals.motif_diversity_score()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifsIntervals.to_bed"><code class="docutils literal notranslate"><span class="pre">EndMotifsIntervals.to_bed()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifsIntervals.to_bedgraph"><code class="docutils literal notranslate"><span class="pre">EndMotifsIntervals.to_bedgraph()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.frag.EndMotifsIntervals.to_tsv"><code class="docutils literal notranslate"><span class="pre">EndMotifsIntervals.to_tsv()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.region_end_motifs"><code class="docutils literal notranslate"><span class="pre">region_end_motifs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.end_motifs"><code class="docutils literal notranslate"><span class="pre">end_motifs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.interval_end_motifs"><code class="docutils literal notranslate"><span class="pre">interval_end_motifs()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cleavage-profile">Cleavage Profile</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.frag.cleavage_profile"><code class="docutils literal notranslate"><span class="pre">cleavage_profile()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#frag-file-utilities">Frag File Utilities</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.utils.filter_bam"><code class="docutils literal notranslate"><span class="pre">filter_bam()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.utils.genome2list"><code class="docutils literal notranslate"><span class="pre">genome2list()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#genome-utilities">Genome Utilities</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.genome.GenomeGaps"><code class="docutils literal notranslate"><span class="pre">GenomeGaps</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.GenomeGaps.b37"><code class="docutils literal notranslate"><span class="pre">GenomeGaps.b37()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.GenomeGaps.get_arm"><code class="docutils literal notranslate"><span class="pre">GenomeGaps.get_arm()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.GenomeGaps.get_contig_gaps"><code class="docutils literal notranslate"><span class="pre">GenomeGaps.get_contig_gaps()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.GenomeGaps.in_tcmere"><code class="docutils literal notranslate"><span class="pre">GenomeGaps.in_tcmere()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.GenomeGaps.overlaps_gap"><code class="docutils literal notranslate"><span class="pre">GenomeGaps.overlaps_gap()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.GenomeGaps.to_bed"><code class="docutils literal notranslate"><span class="pre">GenomeGaps.to_bed()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.GenomeGaps.ucsc_hg19"><code class="docutils literal notranslate"><span class="pre">GenomeGaps.ucsc_hg19()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.genome.ContigGaps"><code class="docutils literal notranslate"><span class="pre">ContigGaps</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.ContigGaps.get_arm"><code class="docutils literal notranslate"><span class="pre">ContigGaps.get_arm()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.ContigGaps.in_gap"><code class="docutils literal notranslate"><span class="pre">ContigGaps.in_gap()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#finaletools.genome.ContigGaps.in_tcmere"><code class="docutils literal notranslate"><span class="pre">ContigGaps.in_tcmere()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.genome.ucsc_hg19_gap_bed"><code class="docutils literal notranslate"><span class="pre">ucsc_hg19_gap_bed()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.genome.b37_gap_bed"><code class="docutils literal notranslate"><span class="pre">b37_gap_bed()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#finaletools.genome.ucsc_hg38_gap_bed"><code class="docutils literal notranslate"><span class="pre">ucsc_hg38_gap_bed()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="releases.html">Releases</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">FinaleTools</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="api">
<h1>API<a class="headerlink" href="#api" title="Permalink to this heading"></a></h1>
<section id="basic-features">
<h2>Basic Features<a class="headerlink" href="#basic-features" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.coverage">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">AlignmentFile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_factor</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.coverage" title="Permalink to this definition"></a></dt>
<dd><p>Return estimated fragment coverage over intervals specified in
<cite>intervals</cite>. Fragments are read from <cite>input_file</cite> which may be
a SAM, BAM, CRAM, or Frag.gz file. Uses an algorithm where the midpoints of
fragments are calculated and coverage is tabulated from the
midpoints that fall into the specified region. Not suitable for
fragments of size approaching interval size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em><em> or </em><em>pysam.AlignmentFile</em>) – SAM, BAM, CRAM, or Frag.gz file containing paired-end fragment
reads or its path. <cite>AlignmentFile</cite> must be opened in read mode.</p></li>
<li><p><strong>interval_file</strong> (<em>str</em>) – BED4 file containing intervals over which to generate coverage
statistics.</p></li>
<li><p><strong>output_file</strong> (<em>string</em><em>, </em><em>optional</em>) – Path for bed file to print coverages to. If output_file = <cite>_</cite>,
results will be printed to stdout.</p></li>
<li><p><strong>scale_factor</strong> (<em>int</em><em>, </em><em>optional</em>) – Amount to multiply coverages by. Default is 10^6.</p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>verbose</strong> (<em>int</em><em> or </em><em>bool</em><em>, </em><em>optional</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>coverage</strong> – Fragment coverage over contig and region.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.frag_length">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">frag_length</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">AlignmentFile</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">TabixFile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intersect_policy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'midpoint'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span></span></span><a class="headerlink" href="#finaletools.frag.frag_length" title="Permalink to this definition"></a></dt>
<dd><p>Return <cite>np.ndarray</cite> containing lengths of fragments in <cite>input_file</cite>
that are above the quality threshold and are proper-paired reads.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em><em> or </em><em>pysam.AlignmentFile</em>) – BAM, SAM, or CRAM file containing paired-end fragment reads or
its path. <cite>AlignmentFile</cite> must be opened in read mode.</p></li>
<li><p><strong>contig</strong> (<em>string</em><em>, </em><em>optional</em>) – Contig or chromosome to get fragments from</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – 0-based left-most coordinate of interval</p></li>
<li><p><strong>stop</strong> (<em>int</em><em>, </em><em>optional</em>) – 1-based right-most coordinate of interval</p></li>
<li><p><strong>intersect_policy</strong> (<em>str</em><em>, </em><em>optional</em>) – Specifies what policy is used to include fragments in the
given interval. Default is “midpoint”. Policies include:
- midpoint: the average of end coordinates of a fragment lies
in the interval.
- any: any part of the fragment is in the interval.</p></li>
<li><p><strong>output_file</strong> (<em>string</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>lengths</strong> – <cite>ndarray</cite> of fragment lengths from file and contig if
specified.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.frag_length_bins">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">frag_length_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">AlignmentFile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig_by_contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">histogram</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intersect_policy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'midpoint'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">ndarray</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#finaletools.frag.frag_length_bins" title="Permalink to this definition"></a></dt>
<dd><p>Takes input_file, computes frag lengths of fragments and returns
two arrays containing bins and counts by size. Optionally prints
data to output as a tab delimited table or histogram.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em><em> or </em><em>AlignmentFile</em>) – </p></li>
<li><p><strong>contig</strong> (<em>str</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>stop</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>bin_size</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>output_file</strong> (<em>str</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>contig_by_contig</strong> (<em>bool</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>histogram</strong> (<em>bool</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>intersect_policy</strong> (<em>str</em><em>, </em><em>optional</em>) – Specifies what policy is used to include fragments in the
given interval. Default is “midpoint”. Policies include:
- midpoint: the average of end coordinates of a fragment lies
in the interval.
- any: any part of the fragment is in the interval.</p></li>
<li><p><strong>workers</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>bins</strong> (<em>ndarray</em>)</p></li>
<li><p><strong>counts</strong> (<em>ndarray</em>)</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="window-protection-score">
<h2>Window Protection Score<a class="headerlink" href="#window-protection-score" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.wps">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">wps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">AlignmentFile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_low</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_high</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">180</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span></span></span><a class="headerlink" href="#finaletools.frag.wps" title="Permalink to this definition"></a></dt>
<dd><p>Return (raw) Windowed Protection Scores as specified in Snyder et al
(2016) over a region [start,stop).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em><em> or </em><em>pysam.AlignmentFile</em>) – BAM, SAM or tabix file containing paired-end fragment reads or its
path. <cite>AlignmentFile</cite> must be opened in read mode.</p></li>
<li><p><strong>contig</strong> (<em>str</em>) – </p></li>
<li><p><strong>start</strong> (<em>int</em>) – </p></li>
<li><p><strong>stop</strong> (<em>int</em>) – </p></li>
<li><p><strong>output_file</strong> (<em>string</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>window_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of window to calculate WPS. Default is k = 120, equivalent
to L-WPS.</p></li>
<li><p><strong>fraction_low</strong> (<em>int</em><em>, </em><em>optional</em>) – Specifies lowest fragment length included in calculation.
Default is 120, equivalent to long fraction.</p></li>
<li><p><strong>fraction_high</strong> (<em>int</em><em>, </em><em>optional</em>) – Specifies highest fragment length included in calculation.
Default is 180, equivalent to long fraction.</p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>workers</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>scores</strong> – np array of shape (n, 2) where column 1 is the coordinate and
column 2 is the score and n is the number of coordinates in
region [start,stop)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.multi_wps">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">multi_wps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AlignmentFile</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_bed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_low</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_high</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">180</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span></span></span><a class="headerlink" href="#finaletools.frag.multi_wps" title="Permalink to this definition"></a></dt>
<dd><p>Function that aggregates WPS over sites in BED file according to the
method described by Snyder et al (2016).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em><em> or </em><em>pysam.AlignmentFile</em>) – BAM, SAM, or tabix file containing paired-end fragment reads or its
path. <cite>AlignmentFile</cite> must be opened in read mode.</p></li>
<li><p><strong>site_bed</strong> (<em>str</em>) – Bed file containing intervals to perform WPS on.</p></li>
<li><p><strong>output_file</strong> (<em>string</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>window_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of window to calculate WPS. Default is k = 120, equivalent
to L-WPS.</p></li>
<li><p><strong>interval_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of each interval specified in the bed file. Should be the
same for every interbal. Default is 5000.</p></li>
<li><p><strong>fraction_low</strong> (<em>int</em><em>, </em><em>optional</em>) – Specifies lowest fragment length included in calculation.
Default is 120, equivalent to long fraction.</p></li>
<li><p><strong>fraction_high</strong> (<em>int</em><em>, </em><em>optional</em>) – Specifies highest fragment length included in calculation.
Default is 120, equivalent to long fraction.</p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>workers</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>scores</strong> – np array of shape (n, 2) where column 1 is the coordinate and
column 2 is the score and n is the number of coordinates in
region [start,stop)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.adjust_wps">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">adjust_wps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">median_window_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">savgol_window_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">21</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">savgol_poly_deg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subtract_edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union(bool,</span> <span class="pre">int)</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.adjust_wps" title="Permalink to this definition"></a></dt>
<dd><p>Adjusts raw WPS data in a BigWig by applying a median filter and
Savitsky-Golay filter (Savitsky and Golay, 1964).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em>) – Path string to a BigWig containing raw WPS data.</p></li>
<li><p><strong>interval_file</strong> (<em>str</em>) – BED format file containing intervals over which WPS was
calculated on.</p></li>
<li><p><strong>output_file</strong> (<em>str</em>) – BigWig file to write adjusted WPS to.</p></li>
<li><p><strong>genome_file</strong> (<em>str</em>) – The genome file for the reference genome that WGS was aligned
to. A tab delimited file where column 1 contains the name of
chromosomes and column 2 contains chromosome length.</p></li>
<li><p><strong>median_window_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of median filter window. Default is 1000.</p></li>
<li><p><strong>savgol_window_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of Savitsky Golay filter window. Default is 21.</p></li>
<li><p><strong>savgol_poly_deg</strong> (<em>int</em><em>, </em><em>optional</em>) – Degree polynomial for Savitsky Golay filter. Default is 2.</p></li>
<li><p><strong>mean</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, a mean filter is used instead of median. Default is
False.</p></li>
<li><p><strong>subtract_edges</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, take the median of the first and last 500 bases in a
window and subtract from the whole interval. Default is False.</p></li>
<li><p><strong>workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of processes to use. Default is 1.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em> or </em><em>int</em><em>, </em><em>optional</em>) – Default is False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.agg_wps">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">agg_wps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">median_window_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.agg_wps" title="Permalink to this definition"></a></dt>
<dd><p>Takes a BigWig containing adjusted WPS scores and an interval BED and
aggregates scores along the intervals.</p>
<p>Note that the median filter trims the ends of each interval by half
of the window size of the filter while adjusting raw WPS data. There
are two way this can be approached in aggregation:
1. supply an interval file containing smaller intervals. e.g. if
you used 5kb intervals for WPS and used a median filter window
of 1kb, supply a BED file with 4kb windows to this function.</p>
<p>2. provide the size of the median filter window in
<cite>median_window_size</cite> along with the original intervals. e.g if
5kb intervals were used for WPS and a 1kb median filter window
was used, supply the 5kb bed file and median filter window size
to this function.</p>
<p>Do not do both of these at once.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em>) – </p></li>
<li><p><strong>interval_file</strong> (<em>str</em>) – </p></li>
<li><p><strong>output_file</strong> (<em>str</em>) – </p></li>
<li><p><strong>median_window_size</strong> (<em>int</em><em>, </em><em>optional</em>) – default is 0</p></li>
<li><p><strong>verbose</strong> (<em>int</em><em> or </em><em>bool</em><em>, </em><em>optional</em>) – default is False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>agg_scores</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>NDArray</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="delfi">
<h2>DELFI<a class="headerlink" href="#delfi" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.delfi">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">delfi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">autosomes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blacklist_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union(str,</span> <span class="pre">GenomeGaps)</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gc_correct</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">merge_bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsample_coverage</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preprocessing</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="headerlink" href="#finaletools.frag.delfi" title="Permalink to this definition"></a></dt>
<dd><p>A function that replicates the methodology of Christiano et al
(2019).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em>) – Path string pointing to a bam file containing PE
fragment reads.</p></li>
<li><p><strong>autosomes</strong> (<em>str</em>) – Path string to a .genome file containing only autosomal
chromosomes</p></li>
<li><p><strong>bins_file</strong> (<em>str</em>) – Path string to a BED file containing 100kb bins for reference
genome of choice. Cristiano et al uses</p></li>
<li><p><strong>reference_file</strong> (<em>str</em>) – Path string to .2bit file.</p></li>
<li><p><strong>blacklist_file</strong> (<em>str</em>) – Path string to BED file containing genome blacklist.</p></li>
<li><p><strong>gap_file</strong> (<em>str</em>) – Path string to a BED4+ file where each interval is a centromere
or telomere. A bed file can be used <strong>only if</strong> the fourth field
for each entry corresponding to a telomere or centromere is
labled “telomere” or “centromere, respectively.</p></li>
<li><p><strong>output_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to output tsv.</p></li>
<li><p><strong>window_size</strong> (<em>int</em>) – Size of non-overlapping windows to cover genome. Default is
5 megabases.</p></li>
<li><p><strong>subsample_coverage</strong> (<em>int</em><em>, </em><em>optional</em>) – The depth at which to subsample the input_bam. Default is 2.</p></li>
<li><p><strong>workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of worker processes to use. Default is 1.</p></li>
<li><p><strong>preprocessing</strong> (<em>bool</em><em>, </em><em>optional</em>) – Christiano et al (2019)</p></li>
<li><p><strong>verbose</strong> (<em>int</em><em> or </em><em>bool</em><em>, </em><em>optional</em>) – Determines how many print statements and loading bars appear in
stdout. Default is False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.delfi_gc_correct">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">delfi_gc_correct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.75</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">it</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.delfi_gc_correct" title="Permalink to this definition"></a></dt>
<dd><p>Helper function that takes window data and performs GC adjustment.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.delfi_merge_bins">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">delfi_merge_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hundred_kb_bins</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gc_corrected</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_chr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.delfi_merge_bins" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="end-motifs">
<h2>End-motifs<a class="headerlink" href="#end-motifs" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifFreqs">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">EndMotifFreqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kmer_frequencies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">20</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.EndMotifFreqs" title="Permalink to this definition"></a></dt>
<dd><p>Class that stores frequencies of end-motif k-mer frequencies and
contains methods to manipulate this data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kmer_frequencies</strong> (<em>Iterable</em>) – A Iterable of tuples, each containing a str representing a k-mer
and a float representing its frequency</p></li>
<li><p><strong>k</strong> (<em>int</em>) – Size of k-mers</p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum mapping quality used. Default is 30.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifFreqs.from_file">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'\t'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#finaletools.frag.EndMotifFreqs" title="finaletools.frag.end_motifs.EndMotifFreqs"><span class="pre">EndMotifFreqs</span></a></span></span><a class="headerlink" href="#finaletools.frag.EndMotifFreqs.from_file" title="Permalink to this definition"></a></dt>
<dd><p>Reads kmer frequency from a two-column tab-delimited file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> (<em>str</em>) – Path string containing path to file.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – Delimiter used in file.</p></li>
<li><p><strong>header</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of lines to ignore at the head of the file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>kmer_freqs</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#finaletools.frag.EndMotifFreqs" title="finaletools.frag.EndMotifFreqs">EndMotifFreqs</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifFreqs.motif_diversity_score">
<span class="sig-name descname"><span class="pre">motif_diversity_score</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#finaletools.frag.EndMotifFreqs.motif_diversity_score" title="Permalink to this definition"></a></dt>
<dd><p>Calculates a motif diversity score (MDS) using normalized
Shannon entropy as described by Jiang et al (2020). This
function is generalized for any k instead of just 4-mers.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifFreqs.to_tsv">
<span class="sig-name descname"><span class="pre">to_tsv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'\t'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.EndMotifFreqs.to_tsv" title="Permalink to this definition"></a></dt>
<dd><p>Prints k-mer frequencies to a tsv</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifsIntervals">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">EndMotifsIntervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">20</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.EndMotifsIntervals" title="Permalink to this definition"></a></dt>
<dd><p>Class that stores frequencies of end-motif k-mers over
user-specified intervals and contains methods to manipulate this
data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>Iterable</em>) – A collection of tuples, each containing a tuple representing
a genomic interval (chrom, 0-based start, 1-based stop) and a
dict that maps kmers to frequencies in the interval.</p></li>
<li><p><strong>k</strong> (<em>int</em>) – Size of k-mers</p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum mapping quality used. Default is 30.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifsIntervals.freq">
<span class="sig-name descname"><span class="pre">freq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kmer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#finaletools.frag.EndMotifsIntervals.freq" title="Permalink to this definition"></a></dt>
<dd><p>Returns a list of intervals and associated frquency for given
kmer. Results are in the form (chrom, 0-based start, 1-based
stop, frequency).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifsIntervals.from_file">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">','</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#finaletools.frag.EndMotifFreqs" title="finaletools.frag.end_motifs.EndMotifFreqs"><span class="pre">EndMotifFreqs</span></a></span></span><a class="headerlink" href="#finaletools.frag.EndMotifsIntervals.from_file" title="Permalink to this definition"></a></dt>
<dd><p>Reads kmer frequency from a tab-delimited file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_path</strong> (<em>str</em>) – Path string containing path to file.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – Delimiter used in file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>kmer_freqs</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#finaletools.frag.EndMotifFreqs" title="finaletools.frag.EndMotifFreqs">EndMotifFreqs</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifsIntervals.mds_bed">
<span class="sig-name descname"><span class="pre">mds_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'\t'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.EndMotifsIntervals.mds_bed" title="Permalink to this definition"></a></dt>
<dd><p>Writes MDS for each interval to a bed/bedgraph file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifsIntervals.motif_diversity_score">
<span class="sig-name descname"><span class="pre">motif_diversity_score</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#finaletools.frag.EndMotifsIntervals.motif_diversity_score" title="Permalink to this definition"></a></dt>
<dd><p>Calculates a motif diversity score (MDS) for each interval using
normalized Shannon entropy as described by Jiang et al (2020). This
function is generalized for any k instead of just 4-mers.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifsIntervals.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kmer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calc_freq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'\t'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.EndMotifsIntervals.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Take frequency of specified kmer and writes to BED.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_file</strong> (<em>str</em>) – File to write frequencies to.</p></li>
<li><p><strong>calc_freq</strong> (<em>bool</em><em>, </em><em>optional</em>) – Calculates frequency of motifs if true. Otherwise, writes counts
for each motif. Default is true.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – Separator for table. Tab-separated by default.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifsIntervals.to_bedgraph">
<span class="sig-name descname"><span class="pre">to_bedgraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kmer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calc_freq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'\t'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.EndMotifsIntervals.to_bedgraph" title="Permalink to this definition"></a></dt>
<dd><p>Take frequency of specified kmer and writes to bedgraph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_file</strong> (<em>str</em>) – File to write frequencies to.</p></li>
<li><p><strong>calc_freq</strong> (<em>bool</em><em>, </em><em>optional</em>) – Calculates frequency of motifs if true. Otherwise, writes counts
for each motif. Default is true.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – Separator for table. Tab-separated by default.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.frag.EndMotifsIntervals.to_tsv">
<span class="sig-name descname"><span class="pre">to_tsv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calc_freq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'\t'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.frag.EndMotifsIntervals.to_tsv" title="Permalink to this definition"></a></dt>
<dd><p>Writes all intervals and associated frquencies to file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_file</strong> (<em>str</em>) – File to write frequencies to.</p></li>
<li><p><strong>calc_freq</strong> (<em>bool</em><em>, </em><em>optional</em>) – Calculates frequency of motifs if true. Otherwise, writes counts
for each motif. Default is true.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – Separator for table. Tab-separated by default.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.region_end_motifs">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">region_end_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refseq_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_low</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_high</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">600</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">both_strands</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#finaletools.frag.region_end_motifs" title="Permalink to this definition"></a></dt>
<dd><p>Function that reads fragments in the specified region from a BAM,
SAM, or tabix indexed file and returns the 5’ k-mer (default is
4-mer) end motif counts as a dictionary. This function
reproduces the methodology of Zhou et al (2023).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em>) – Path of SAM, BAM, CRAM, or Frag.gz containing pair-end reads.</p></li>
<li><p><strong>contig</strong> (<em>str</em>) – Name of contig or chromosome for region.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – 0-based start coordinate.</p></li>
<li><p><strong>stop</strong> (<em>int</em>) – 1-based end coordinate.</p></li>
<li><p><strong>refseq_file</strong> (<em>str</em>) – 2bit file with reference sequence <cite>input_file</cite> was aligned to.</p></li>
<li><p><strong>k</strong> (<em>int</em><em>, </em><em>optional</em>) – Length of end motif kmer. Default is 4.</p></li>
<li><p><strong>fraction_low</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum fragment length.</p></li>
<li><p><strong>fraction_high</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum fragment length.</p></li>
<li><p><strong>both_strands</strong> (<em>bool</em><em>, </em><em>optional</em>) – Choose whether to use forward 5’ ends only or use 5’ ends for
both ends of PE reads.</p></li>
<li><p><strong>output_file</strong> (<em>None</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em> or </em><em>int</em><em>, </em><em>optional</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>end_motif_freq</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.end_motifs">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">end_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refseq_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_low</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_high</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">600</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">both_strands</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#finaletools.frag.EndMotifFreqs" title="finaletools.frag.end_motifs.EndMotifFreqs"><span class="pre">EndMotifFreqs</span></a></span></span><a class="headerlink" href="#finaletools.frag.end_motifs" title="Permalink to this definition"></a></dt>
<dd><p>Function that reads fragments from a BAM, SAM, or tabix indexed
file and returns the 5’ k-mer (default is 4-mer) end motif
frequencies as a dictionary. Optionally writes data to a tsv. This
function reproduces the methodology of Zhou et al (2023).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em>) – SAM, BAM, CRAM, or Frag.gz file with paired-end reads.</p></li>
<li><p><strong>refseq_file</strong> (<em>str</em>) – 2bit file with sequence of reference genome input_file is
aligned to.</p></li>
<li><p><strong>k</strong> (<em>int</em><em>, </em><em>optional</em>) – Length of end motif kmer. Default is 4.</p></li>
<li><p><strong>output_file</strong> (<em>None</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – File path to write results to. Either tsv or csv.</p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum MAPQ to filter.</p></li>
<li><p><strong>workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of worker processes.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em> or </em><em>int</em><em>, </em><em>optional</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>end_motif_freq</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#finaletools.frag.EndMotifFreqs" title="finaletools.frag.EndMotifFreqs">EndMotifFreqs</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.interval_end_motifs">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">interval_end_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refseq_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intervals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_low</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_high</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">600</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">both_strands</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#finaletools.frag.EndMotifFreqs" title="finaletools.frag.end_motifs.EndMotifFreqs"><span class="pre">EndMotifFreqs</span></a></span></span><a class="headerlink" href="#finaletools.frag.interval_end_motifs" title="Permalink to this definition"></a></dt>
<dd><p>Function that reads fragments from a BAM, SAM, or tabix indexed
file and user-specified intervals and returns the 5’ k-mer
(default is 4-mer) end motif. Optionally writes data to a tsv.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_file</strong> (<em>str</em>) – Path of SAM, BAM, CRAM, or Frag.gz containing pair-end reads.</p></li>
<li><p><strong>refseq_file</strong> (<em>str</em>) – Path of 2bit file for reference genome that reads are aligned to.</p></li>
<li><p><strong>intervals</strong> (<em>str</em><em> or </em><em>tuple</em>) – Path of BED file containing intervals or list of tuples
(chrom, start, stop).</p></li>
<li><p><strong>k</strong> (<em>int</em><em>, </em><em>optional</em>) – Length of end motif kmer. Default is 4.</p></li>
<li><p><strong>output_file</strong> (<em>None</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – File path to write results to. Either tsv or csv.</p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum MAPQ to filter.</p></li>
<li><p><strong>workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of worker processes.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em> or </em><em>int</em><em>, </em><em>optional</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>end_motif_freq</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>EndMotifIntervals</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="cleavage-profile">
<h2>Cleavage Profile<a class="headerlink" href="#cleavage-profile" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="finaletools.frag.cleavage_profile">
<span class="sig-prename descclassname"><span class="pre">finaletools.frag.</span></span><span class="sig-name descname"><span class="pre">cleavage_profile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_low</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_high</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10000000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">ndarray</span></span></span><a class="headerlink" href="#finaletools.frag.cleavage_profile" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="frag-file-utilities">
<h2>Frag File Utilities<a class="headerlink" href="#frag-file-utilities" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="finaletools.utils.filter_bam">
<span class="sig-prename descclassname"><span class="pre">finaletools.utils.</span></span><span class="sig-name descname"><span class="pre">filter_bam</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_high</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_low</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.utils.filter_bam" title="Permalink to this definition"></a></dt>
<dd><p>Accepts the path to a BAM file and creates a bam file where all
reads are read1 in a proper pair, exceed the specified quality
threshold, do not intersect a region in the given blacklist
file, and intersects with a region in the region bed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_bam</strong> (<em>str</em>) – Path string or AlignmentFile pointing to the BAM file to be
filtered.</p></li>
<li><p><strong>region_file</strong> (<em>str</em><em>, </em><em>option</em>) – </p></li>
<li><p><strong>output_file</strong> (<em>str</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>workers</strong> (<em>int</em><em>, </em><em>optional</em>) – </p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>output_file</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.utils.genome2list">
<span class="sig-prename descclassname"><span class="pre">finaletools.utils.</span></span><span class="sig-name descname"><span class="pre">genome2list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="headerlink" href="#finaletools.utils.genome2list" title="Permalink to this definition"></a></dt>
<dd><p>Reads a GENOME text file into a list of tuples (chrom, length)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>genome_file</strong> (<em>str</em>) – String containing path to GENOME format file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>chroms</strong> – List of tuples containing chrom/contig names and lengths</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="genome-utilities">
<h2>Genome Utilities<a class="headerlink" href="#genome-utilities" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="finaletools.genome.GenomeGaps">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">finaletools.genome.</span></span><span class="sig-name descname"><span class="pre">GenomeGaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gaps_bed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.genome.GenomeGaps" title="Permalink to this definition"></a></dt>
<dd><p>Reads telomere, centromere, and short_arm intervals from a bed file
or generates these intervals from UCSC gap and centromere tracks for
hg19 and hg38.</p>
<dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.GenomeGaps.b37">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">b37</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.genome.GenomeGaps.b37" title="Permalink to this definition"></a></dt>
<dd><p>Creates a GenomeGaps for the Broad Institute GRCh37 reference
genome i.e b37. This reference genome is also based on GRCh37,
but differs from the UCSC hg19 reference in a few ways,
including the absence of the ‘chr’ prefix. We generate this
GenomeGap using an ad hoc method where we take the UCSC hg19
gap track and drop ‘chr’ from the chromosome names. Because
there are other differences between hg19 and b37, this is not
a perfect solution.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gaps</strong> – GenomeGaps for the b37 reference genome.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#finaletools.genome.GenomeGaps" title="finaletools.genome.GenomeGaps">GenomeGaps</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.GenomeGaps.get_arm">
<span class="sig-name descname"><span class="pre">get_arm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#finaletools.genome.GenomeGaps.get_arm" title="Permalink to this definition"></a></dt>
<dd><p>Returns the chromosome arm the interval is in. If in
the short arm of an acrocentric chromosome or intersects a
centromere, returns an empty string.</p>
<dl class="simple">
<dt>contig<span class="classifier">str</span></dt><dd><p>Chromosome of interval.</p>
</dd>
<dt>start<span class="classifier">int</span></dt><dd><p>Start of interval.</p>
</dd>
<dt>stop<span class="classifier">int</span></dt><dd><p>End of interval.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>arm</strong> – Arm that interval is in.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – Raised for invalid coordinates</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.GenomeGaps.get_contig_gaps">
<span class="sig-name descname"><span class="pre">get_contig_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#finaletools.genome.ContigGaps" title="finaletools.genome.gaps.ContigGaps"><span class="pre">ContigGaps</span></a></span></span><a class="headerlink" href="#finaletools.genome.GenomeGaps.get_contig_gaps" title="Permalink to this definition"></a></dt>
<dd><p>Creates a ContigGaps for the specified chromosome</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>contig</strong> (<em>str</em>) – Chromosome to make ContigGaps for</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Contains centromere and telomere intervals for chromosome</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#finaletools.genome.ContigGaps" title="finaletools.genome.ContigGaps">ContigGaps</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.GenomeGaps.in_tcmere">
<span class="sig-name descname"><span class="pre">in_tcmere</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#finaletools.genome.GenomeGaps.in_tcmere" title="Permalink to this definition"></a></dt>
<dd><p>Checks if specified interval is in a centromere or telomere</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contig</strong> (<em>str</em>) – Chromosome name</p></li>
<li><p><strong>start</strong> (<em>int</em>) – Start of interval</p></li>
<li><p><strong>stop</strong> (<em>int</em>) – End of interval</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>in_telomere_or_centromere</strong> – True if in a centromere or telomere</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.GenomeGaps.overlaps_gap">
<span class="sig-name descname"><span class="pre">overlaps_gap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#finaletools.genome.GenomeGaps.overlaps_gap" title="Permalink to this definition"></a></dt>
<dd><p>Checks if specified interval overlaps a gap interval</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contig</strong> (<em>str</em>) – Chromosome name</p></li>
<li><p><strong>start</strong> (<em>int</em>) – Start of interval</p></li>
<li><p><strong>stop</strong> (<em>int</em>) – End of interval</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>in_telomere_or_centromere</strong> – True if in a centromere or telomere</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.GenomeGaps.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.genome.GenomeGaps.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Prints gap intervals in GenomeGaps to a BED4 file where the name
is the type of gap interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_file</strong> (<em>str</em>) – File to write to. Optionally gzipped. If output_file == ‘-‘,
results will be writted to stdout.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.GenomeGaps.ucsc_hg19">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ucsc_hg19</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.genome.GenomeGaps.ucsc_hg19" title="Permalink to this definition"></a></dt>
<dd><p>Creates a GenomeGaps for the UCSC hg19 reference genome. This
sequences uses chromosome names that start with ‘chr’ and is
based on a version of the GRCh37 reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gaps</strong> – GenomeGaps for the UCSC hg19 reference genome.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#finaletools.genome.GenomeGaps" title="finaletools.genome.GenomeGaps">GenomeGaps</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="finaletools.genome.ContigGaps">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">finaletools.genome.</span></span><span class="sig-name descname"><span class="pre">ContigGaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centromere</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">telomeres</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">has_short_arm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.genome.ContigGaps" title="Permalink to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.ContigGaps.get_arm">
<span class="sig-name descname"><span class="pre">get_arm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.genome.ContigGaps.get_arm" title="Permalink to this definition"></a></dt>
<dd><p>Returns name of chromosome arm the interval is in. Returns
“NOARM” if in a centromere, telomere, or short arm of an
acrocentric chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>int</em>) – Start of interval.</p></li>
<li><p><strong>stop</strong> (<em>int</em>) – End of interval.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Name of the chromosome arm.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – Raised if invalid coordinates are given.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.ContigGaps.in_gap">
<span class="sig-name descname"><span class="pre">in_gap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#finaletools.genome.ContigGaps.in_gap" title="Permalink to this definition"></a></dt>
<dd><p>Checks if specified interval is in a gap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>int</em>) – Start of interval.</p></li>
<li><p><strong>stop</strong> (<em>int</em>) – End of Interval.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if there is an overlap.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="finaletools.genome.ContigGaps.in_tcmere">
<span class="sig-name descname"><span class="pre">in_tcmere</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#finaletools.genome.ContigGaps.in_tcmere" title="Permalink to this definition"></a></dt>
<dd><p>Checks if specified interval is in a centromere or telomere.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>int</em>) – Start of interval.</p></li>
<li><p><strong>stop</strong> (<em>int</em>) – End of Interval.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if there is an overlap.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.genome.ucsc_hg19_gap_bed">
<span class="sig-prename descclassname"><span class="pre">finaletools.genome.</span></span><span class="sig-name descname"><span class="pre">ucsc_hg19_gap_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.genome.ucsc_hg19_gap_bed" title="Permalink to this definition"></a></dt>
<dd><p>Creates BED4 of centromeres, telomeres, and short arms for the UCSC
hg19 reference sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_file</strong> (<em>str</em>) – Output path</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.genome.b37_gap_bed">
<span class="sig-prename descclassname"><span class="pre">finaletools.genome.</span></span><span class="sig-name descname"><span class="pre">b37_gap_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.genome.b37_gap_bed" title="Permalink to this definition"></a></dt>
<dd><p>Creates BED4 of centromeres, telomeres, and short arms for the Broad
Institute GRCh37 (b37) reference sequence. Also useful for
files aligned to human_g1k_v37 (1000 Genomes Project).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_file</strong> (<em>str</em>) – Output path</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="finaletools.genome.ucsc_hg38_gap_bed">
<span class="sig-prename descclassname"><span class="pre">finaletools.genome.</span></span><span class="sig-name descname"><span class="pre">ucsc_hg38_gap_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#finaletools.genome.ucsc_hg38_gap_bed" title="Permalink to this definition"></a></dt>
<dd><p>Creates BED4 of centromeres, telomeres, and short arms for the UCSC
hg38 reference sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_file</strong> (<em>str</em>) – Output path</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="cli.html" class="btn btn-neutral float-left" title="CLI" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="releases.html" class="btn btn-neutral float-right" title="Releases" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, James Li, Yaping Liu.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>