
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>CLI &#8212; FinaleToolkit  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=a746c00c" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'documentation/cli_reference/index';</script>
    <link rel="icon" href="../../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="API" href="../api_reference/index.html" />
    <link rel="prev" title="Help" href="../user_guide/help.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/finaletoolkit_logo_rounded.png" class="logo__image only-light" alt=""/>
    <script>document.write(`<img src="../../_static/finaletoolkit_logo_rounded.png" class="logo__image only-dark" alt=""/>`);</script>
  
  
    <p class="title logo__title">FinaleToolkit</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    FinaleToolkit Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide/index.html">
    User Guide
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    CLI
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../api_reference/index.html">
    API
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../index.html">
    FinaleToolkit Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide/index.html">
    User Guide
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    CLI
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../api_reference/index.html">
    API
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"></div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../index.html" class="nav-link">FinaleToolkit Documentation</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">CLI</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="cli">
<h1>CLI<a class="headerlink" href="#cli" title="Link to this heading">#</a></h1>
<p><p>FinaleToolkit is a package and standalone program to extract fragmentation features of cell-free DNA from paired-end sequencing data.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">finaletoolkit</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                     <span class="p">{</span><span class="n">coverage</span><span class="p">,</span><span class="n">frag</span><span class="o">-</span><span class="n">length</span><span class="o">-</span><span class="n">bins</span><span class="p">,</span><span class="n">frag</span><span class="o">-</span><span class="n">length</span><span class="o">-</span><span class="n">intervals</span><span class="p">,</span><span class="n">cleavage</span><span class="o">-</span><span class="n">profile</span><span class="p">,</span><span class="n">wps</span><span class="p">,</span><span class="n">adjust</span><span class="o">-</span><span class="n">wps</span><span class="p">,</span><span class="n">delfi</span><span class="p">,</span><span class="n">delfi</span><span class="o">-</span><span class="n">gc</span><span class="o">-</span><span class="n">correct</span><span class="p">,</span><span class="n">end</span><span class="o">-</span><span class="n">motifs</span><span class="p">,</span><span class="n">interval</span><span class="o">-</span><span class="n">end</span><span class="o">-</span><span class="n">motifs</span><span class="p">,</span><span class="n">mds</span><span class="p">,</span><span class="n">interval</span><span class="o">-</span><span class="n">mds</span><span class="p">,</span><span class="nb">filter</span><span class="o">-</span><span class="n">bam</span><span class="p">,</span><span class="n">agg</span><span class="o">-</span><span class="n">bw</span><span class="p">,</span><span class="n">gap</span><span class="o">-</span><span class="n">bed</span><span class="p">}</span>
                     <span class="o">...</span>
</pre></div>
</div>
<section id="named-arguments">
<h2>Named Arguments<a class="headerlink" href="#named-arguments" title="Link to this heading">#</a></h2>
<dl class="option-list">
<dt><kbd>-v, --version</kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
</dl>
</section>
<section id="Sub-commands">
<h2>Sub-commands<a class="headerlink" href="#Sub-commands" title="Link to this heading">#</a></h2>
<section id="coverage">
<h3>coverage<a class="headerlink" href="#coverage" title="Link to this heading">#</a></h3>
<p>Calculates fragmentation coverage over intervals defined in a BED file based on alignment data from a BAM/SAM/CRAM/Fragment file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">coverage</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SCALE_FACTOR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUALITY_THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WORKERS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                       <span class="n">input_file</span> <span class="n">interval_file</span>
</pre></div>
</div>
<section id="positional-arguments">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>Path to a BAM/SAM/CRAM/Fragment file containing fragment data.</p>
</dd>
<dt><kbd>interval_file</kbd></dt>
<dd><p>Path to a BED file containing intervals to calculate coverage over.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat1">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat1" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-o, --output_file</kbd></dt>
<dd><p>A BED file containing coverage values over the intervals specified in interval file.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-s, --scale-factor</kbd></dt>
<dd><p>Scale factor for coverage values.</p>
<p>Default: 1000000.0</p>
</dd>
<dt><kbd>-q, --quality_threshold</kbd></dt>
<dd><p>Minimum mapping quality threshold.</p>
<p>Default: 30</p>
</dd>
<dt><kbd>-w, --workers</kbd></dt>
<dd><p>Number of worker processes.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
<p>Default: False</p>
</dd>
</dl>
</section>
</section>
<section id="frag-length-bins">
<h3>frag-length-bins<a class="headerlink" href="#frag-length-bins" title="Link to this heading">#</a></h3>
<p>Retrieves fragment lengths grouped in bins given a BAM/SAM/CRAM/Fragment file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">frag</span><span class="o">-</span><span class="n">length</span><span class="o">-</span><span class="n">bins</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">CONTIG</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">S</span> <span class="n">START</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="p">{</span><span class="n">midpoint</span><span class="p">,</span><span class="nb">any</span><span class="p">}]</span> <span class="p">[</span><span class="o">-</span><span class="n">E</span> <span class="n">STOP</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">bin</span><span class="o">-</span><span class="n">size</span> <span class="n">BIN_SIZE</span><span class="p">]</span>
                               <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">contig</span><span class="o">-</span><span class="n">by</span><span class="o">-</span><span class="n">contig</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">histogram</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUALITY_THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                               <span class="n">input_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat1">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat1" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>Path to a BAM/SAM/CRAM/Fragment file containing fragment data.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat2">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat2" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-c, --contig</kbd></dt>
<dd><p>Specify the contig or chromosome to select fragments from. (Required if using –start or –stop.)</p>
</dd>
<dt><kbd>-S, --start</kbd></dt>
<dd><p>Specify the 0-based left-most coordinate of the interval to select fragments from. (Must also specify –contig.)</p>
</dd>
<dt><kbd>-p, --intersect_policy</kbd></dt>
<dd><p>Possible choices: midpoint, any</p>
<p>Specifies what policy is used to include fragments in the given interval. See User Guide for more information.</p>
<p>Default: “midpoint”</p>
</dd>
<dt><kbd>-E, --stop</kbd></dt>
<dd><p>Specify the 1-based right-most coordinate of the interval to select fragments from. (Must also specify –contig.)</p>
</dd>
<dt><kbd>--bin-size</kbd></dt>
<dd><p>Specify the size of the bins to group fragment lengths into.</p>
</dd>
<dt><kbd>-o, --output_file</kbd></dt>
<dd><p>A .TSV file containing containing fragment lengths binned according to the specified bin size.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>--contig-by-contig</kbd></dt>
<dd><p>Placeholder, not implemented.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--histogram</kbd></dt>
<dd><p>Enable histogram mode to display histogram in terminal.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-q, --quality_threshold</kbd></dt>
<dd><p>Minimum mapping quality threshold.</p>
<p>Default: 30</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
<p>Default: 0</p>
</dd>
</dl>
</section>
</section>
<section id="frag-length-intervals">
<h3>frag-length-intervals<a class="headerlink" href="#frag-length-intervals" title="Link to this heading">#</a></h3>
<p>Retrieves fragment length summary statistics over intervals defined in a BED file based on alignment data from a BAM/SAM/CRAM/Fragment file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">frag</span><span class="o">-</span><span class="n">length</span><span class="o">-</span><span class="n">intervals</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="p">{</span><span class="n">midpoint</span><span class="p">,</span><span class="nb">any</span><span class="p">}]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUALITY_THRESHOLD</span><span class="p">]</span>
                                    <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WORKERS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                                    <span class="n">input_file</span> <span class="n">interval_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat2">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat2" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>Path to a BAM/SAM/CRAM/Fragment file containing fragment data.</p>
</dd>
<dt><kbd>interval_file</kbd></dt>
<dd><p>Path to a BED file containing intervals to retrieve fragment length summary statistics over.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat3">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat3" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-p, --intersect_policy</kbd></dt>
<dd><p>Possible choices: midpoint, any</p>
<p>Specifies what policy is used to include fragments in the given interval. See User Guide for more information.</p>
<p>Default: “midpoint”</p>
</dd>
<dt><kbd>-o, --output-file</kbd></dt>
<dd><p>A BED file containing fragment length summary statistics (mean, median, st. dev, min, max) over the intervals specified in the interval file.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-q, --quality-threshold</kbd></dt>
<dd><p>Minimum mapping quality threshold.</p>
<p>Default: 30</p>
</dd>
<dt><kbd>-w, --workers</kbd></dt>
<dd><p>Number of worker processes.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
<p>Default: 0</p>
</dd>
</dl>
</section>
</section>
<section id="cleavage-profile">
<h3>cleavage-profile<a class="headerlink" href="#cleavage-profile" title="Link to this heading">#</a></h3>
<p>Calculates cleavage proportion over intervals defined in a BED file based on alignment data from a BAM/SAM/CRAM/Fragment file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">cleavage</span><span class="o">-</span><span class="n">profile</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">lo</span> <span class="n">FRACTION_LOW</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">hi</span> <span class="n">FRACTION_HIGH</span><span class="p">]</span>
                               <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUALITY_THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">LEFT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">r</span> <span class="n">RIGHT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WORKERS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                               <span class="n">input_file</span> <span class="n">interval_file</span> <span class="n">chrom_sizes</span>
</pre></div>
</div>
<section id="positional-arguments_repeat3">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat3" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>Path to a BAM/SAM/CRAM/Fragment file containing fragment data.</p>
</dd>
<dt><kbd>interval_file</kbd></dt>
<dd><p>Path to a BED file containing intervals to calculates cleavage proportion over.</p>
</dd>
<dt><kbd>chrom_sizes</kbd></dt>
<dd><p>A .chrom.sizes file containing chromosome sizes.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat4">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat4" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-o, --output_file</kbd></dt>
<dd><p>A bigWig file containing the cleavage proportion results over the intervals specified in interval file.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-lo, --fraction_low</kbd></dt>
<dd><p>Minimum length for a fragment to be included in cleavage proportion calculation.</p>
<p>Default: 120</p>
</dd>
<dt><kbd>-hi, --fraction_high</kbd></dt>
<dd><p>Maximum length for a fragment to be included in cleavage proportion calculation.</p>
<p>Default: 180</p>
</dd>
<dt><kbd>-q, --quality-threshold</kbd></dt>
<dd><p>Minimum mapping quality threshold.</p>
<p>Default: 20</p>
</dd>
<dt><kbd>-l, --left</kbd></dt>
<dd><p>Number of base pairs to subtract from start coordinate to create interval. Useful when dealing with BED files with only CpG coordinates.</p>
<p>Default: 0</p>
</dd>
<dt><kbd>-r, --right</kbd></dt>
<dd><p>Number of base pairs to add to stop coordinate to create interval. Useful when dealing with BED files with only CpG coordinates.</p>
<p>Default: 0</p>
</dd>
<dt><kbd>-w, --workers</kbd></dt>
<dd><p>Number of worker processes.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
<p>Default: 0</p>
</dd>
</dl>
</section>
</section>
<section id="wps">
<h3>wps<a class="headerlink" href="#wps" title="Link to this heading">#</a></h3>
<p>Calculates Windowed Protection Score (WPS) over intervals defined in a BED file based on alignment data from a BAM/SAM/CRAM/Fragment file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">wps</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INTERVAL_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">W</span> <span class="n">WINDOW_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">lo</span> <span class="n">FRACTION_LOW</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">-</span><span class="n">hi</span> <span class="n">FRACTION_HIGH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUALITY_THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WORKERS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                  <span class="n">input_file</span> <span class="n">site_bed</span>
</pre></div>
</div>
<section id="positional-arguments_repeat4">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat4" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>Path to a BAM/SAM/CRAM/Fragment file containing fragment data.</p>
</dd>
<dt><kbd>site_bed</kbd></dt>
<dd><p>Path to a BED file containing intervals to calculate WPS over.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat5">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat5" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-o, --output_file</kbd></dt>
<dd><p>A bigWig file containing the WPS results over the intervals specified in interval file.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-i, --interval_size</kbd></dt>
<dd><p>Size in bp of each interval in the interval file.</p>
<p>Default: 5000</p>
</dd>
<dt><kbd>-W, --window_size</kbd></dt>
<dd><p>Size of the sliding window used to calculate WPS scores.</p>
<p>Default: 120</p>
</dd>
<dt><kbd>-lo, --fraction_low</kbd></dt>
<dd><p>Minimum length for a fragment to be included in WPS calculation.</p>
<p>Default: 120</p>
</dd>
<dt><kbd>-hi, --fraction_high</kbd></dt>
<dd><p>Maximum length for a fragment to be included in WPS calculation.</p>
<p>Default: 180</p>
</dd>
<dt><kbd>-q, --quality_threshold</kbd></dt>
<dd><p>Minimum mapping quality threshold.</p>
<p>Default: 30</p>
</dd>
<dt><kbd>-w, --workers</kbd></dt>
<dd><p>Number of worker processes.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
<p>Default: 0</p>
</dd>
</dl>
</section>
</section>
<section id="adjust-wps">
<h3>adjust-wps<a class="headerlink" href="#adjust-wps" title="Link to this heading">#</a></h3>
<p>Adjusts raw Windowed Protection Score (WPS) by applying a median filter and Savitsky-Golay filter.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">adjust</span><span class="o">-</span><span class="n">wps</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INTERVAL_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span> <span class="n">MEDIAN_WINDOW_SIZE</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SAVGOL_WINDOW_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="n">SAVGOL_POLY_DEG</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WORKERS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">mean</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">subtract</span><span class="o">-</span><span class="n">edges</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                         <span class="n">input_file</span> <span class="n">interval_file</span> <span class="n">genome_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat5">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat5" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>A bigWig file containing the WPS results over the intervals specified in interval file.</p>
</dd>
<dt><kbd>interval_file</kbd></dt>
<dd><p>Path to a BED file containing intervals to WPS was calculated over.</p>
</dd>
<dt><kbd>genome_file</kbd></dt>
<dd><p>A .chrom.sizes file containing chromosome sizes.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat6">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat6" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-o, --output-file</kbd></dt>
<dd><p>A bigWig file containing the adjusted WPS results over the intervals specified in interval file.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-i, --interval_size</kbd></dt>
<dd><p>Size in bp of each interval in the interval file.</p>
<p>Default: 5000</p>
</dd>
<dt><kbd>-m, --median-window-size</kbd></dt>
<dd><p>Size of the median filter window used to adjust WPS scores.</p>
<p>Default: 1000</p>
</dd>
<dt><kbd>-s, --savgol-window-size</kbd></dt>
<dd><p>Size of the Savitsky-Golay filter window used to adjust WPS scores.</p>
<p>Default: 21</p>
</dd>
<dt><kbd>-p, --savgol-poly-deg</kbd></dt>
<dd><p>Degree polynomial for Savitsky-Golay filter.</p>
<p>Default: 2</p>
</dd>
<dt><kbd>-w, --workers</kbd></dt>
<dd><p>Number of worker processes.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--mean</kbd></dt>
<dd><p>A mean filter is used instead of median.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--subtract-edges</kbd></dt>
<dd><p>Take the median of the first and last 500 bases in a window and subtract from the whole interval.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
</dd>
</dl>
</section>
</section>
<section id="delfi">
<h3>delfi<a class="headerlink" href="#delfi" title="Link to this heading">#</a></h3>
<p>Calculates DELFI featrues over genome, returning information about (GC-corrected) short fragments, long fragments, DELFI ratio, and total fragments. NOTE: Due to some ad hoc implementation details, currently the only accepted reference genome is hg19.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">delfi</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BLACKLIST_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span> <span class="n">GAP_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">G</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">R</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">M</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUALITY_THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WORKERS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                    <span class="n">input_file</span> <span class="n">autosomes</span> <span class="n">reference_file</span> <span class="n">bins_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat6">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat6" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>Path to a BAM/SAM/CRAM/Fragment file containing fragment data.</p>
</dd>
<dt><kbd>autosomes</kbd></dt>
<dd><p>Tab-delimited file containing (1) autosome name and (2) integer length of chromosome in base pairs.</p>
</dd>
<dt><kbd>reference_file</kbd></dt>
<dd><p>The .2bit file for the associate reference genome sequence used during alignment.</p>
</dd>
<dt><kbd>bins_file</kbd></dt>
<dd><p>A BED file containing bins over which to calculate DELFI. To replicate Cristiano et al.’s methodology, use 100kb bins over human autosomes.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat7">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat7" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-b, --blacklist-file</kbd></dt>
<dd><p>BED file containing regions to ignore when calculating DELFI.</p>
</dd>
<dt><kbd>-g, --gap-file</kbd></dt>
<dd><p>BED4 format file containing columns for “chrom”, “start”, “stop”, and “type”. The “type” column should denote whether the entry corresponds to a “centromere”, “telomere”, or “short arm”, and entries not falling into these categories are ignored. This information corresponds to the “gap” track for hg19 in the UCSC Genome Browser.</p>
</dd>
<dt><kbd>-o, --output-file</kbd></dt>
<dd><p>BED, bed.gz, TSV, or CSV file to write DELFI data to. If “-”, writes to stdout.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-G, --no-gc-correct</kbd></dt>
<dd><p>Skip GC correction.</p>
<p>Default: True</p>
</dd>
<dt><kbd>-R, --keep-nocov</kbd></dt>
<dd><p>Skip removal two regions in hg19 with no coverage. Use this flag when not using hg19 human reference genome.</p>
<p>Default: True</p>
</dd>
<dt><kbd>-M, --no-merge-bins</kbd></dt>
<dd><p>Keep 100kb bins and do not merge to 5Mb size.</p>
<p>Default: True</p>
</dd>
<dt><kbd>-q, --quality-threshold</kbd></dt>
<dd><p>Minimum mapping quality threshold.</p>
<p>Default: 30</p>
</dd>
<dt><kbd>-w, --workers</kbd></dt>
<dd><p>Number of worker processes.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
<p>Default: 0</p>
</dd>
</dl>
</section>
</section>
<section id="delfi-gc-correct">
<h3>delfi-gc-correct<a class="headerlink" href="#delfi-gc-correct" title="Link to this heading">#</a></h3>
<p>Performs gc-correction on raw delfi data.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">delfi</span><span class="o">-</span><span class="n">gc</span><span class="o">-</span><span class="n">correct</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">header</span><span class="o">-</span><span class="n">lines</span> <span class="n">HEADER_LINES</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="n">input_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat7">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat7" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>BED file containing raw DELFI data. Raw DELFI data should only have columns for “contig”, “start”, “stop”, “arm”, “short”, “long”, “gc”, “num_frags”, “ratio”.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat8">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat8" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-o, --output-file</kbd></dt>
<dd><p>BED to print GC-corrected DELFI fractions. If “-”, will write to stdout.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>--header-lines</kbd></dt>
<dd><p>Number of header lines in BED.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
</dd>
</dl>
</section>
</section>
<section id="end-motifs">
<h3>end-motifs<a class="headerlink" href="#end-motifs" title="Link to this heading">#</a></h3>
<p>Measures frequency of k-mer 5’ end motifs.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">end</span><span class="o">-</span><span class="n">motifs</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">k</span> <span class="n">K</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUALITY_THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WORKERS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                         <span class="n">input_file</span> <span class="n">refseq_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat8">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat8" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>Path to a BAM/SAM/CRAM/Fragment file containing fragment data.</p>
</dd>
<dt><kbd>refseq_file</kbd></dt>
<dd><p>The .2bit file for the associate reference genome sequence used during alignment.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat9">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat9" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-k</kbd></dt>
<dd><p>Length of k-mer.</p>
<p>Default: 4</p>
</dd>
<dt><kbd>-o, --output-file</kbd></dt>
<dd><p>TSV to print k-mer frequencies. If “-”, will write to stdout.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-q, --quality-threshold</kbd></dt>
<dd><p>Minimum mapping quality threshold.</p>
<p>Default: 20</p>
</dd>
<dt><kbd>-w, --workers</kbd></dt>
<dd><p>Number of worker processes.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
<p>Default: 0</p>
</dd>
</dl>
</section>
</section>
<section id="interval-end-motifs">
<h3>interval-end-motifs<a class="headerlink" href="#interval-end-motifs" title="Link to this heading">#</a></h3>
<p>Measures frequency of k-mer 5’ end motifs in each region specified in a BED file and writes data into a table.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">interval</span><span class="o">-</span><span class="n">end</span><span class="o">-</span><span class="n">motifs</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">k</span> <span class="n">K</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">lo</span> <span class="n">FRACTION_LOW</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">hi</span> <span class="n">FRACTION_HIGH</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span>
                                  <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUALITY_THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WORKERS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                                  <span class="n">input_file</span> <span class="n">refseq_file</span> <span class="n">intervals</span>
</pre></div>
</div>
<section id="positional-arguments_repeat9">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat9" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>Path to a BAM/SAM/CRAM/Fragment file containing fragment data.</p>
</dd>
<dt><kbd>refseq_file</kbd></dt>
<dd><p>The .2bit file for the associate reference genome sequence used during alignment.</p>
</dd>
<dt><kbd>intervals</kbd></dt>
<dd><p>Path to a BED file containing intervals to retrieve end motif frequencies over.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat10">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat10" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-k</kbd></dt>
<dd><p>Length of k-mer.</p>
<p>Default: 4</p>
</dd>
<dt><kbd>-lo, --fraction-low</kbd></dt>
<dd><p>Minimum length for a fragment to be included in end motif frequency.</p>
<p>Default: 10</p>
</dd>
<dt><kbd>-hi, --fraction-high</kbd></dt>
<dd><p>Maximum length for a fragment to be included in end motif frequency.</p>
<p>Default: 600</p>
</dd>
<dt><kbd>-o, --output-file</kbd></dt>
<dd><p>Path to TSV or CSV file to write end motif frequencies to.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-q, --quality-threshold</kbd></dt>
<dd><p>Minimum mapping quality threshold.</p>
<p>Default: 20</p>
</dd>
<dt><kbd>-w, --workers</kbd></dt>
<dd><p>Number of worker processes.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
<p>Default: 0</p>
</dd>
</dl>
</section>
</section>
<section id="mds">
<h3>mds<a class="headerlink" href="#mds" title="Link to this heading">#</a></h3>
<p>Reads k-mer frequencies from a file and calculates a motif diversity score (MDS) using normalized Shannon entropy as described by Jiang et al (2020).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">mds</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEP</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">header</span> <span class="n">HEADER</span><span class="p">]</span> <span class="p">[</span><span class="n">file_path</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments_repeat10">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat10" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>file_path</kbd></dt>
<dd><p>Tab-delimited or similar file containing one column for all k-mers a one column for frequency. Reads from stdin by default.</p>
<p>Default: “-”</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat11">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat11" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-s, --sep</kbd></dt>
<dd><p>Separator used in tabular file.</p>
<p>Default: “      “</p>
</dd>
<dt><kbd>--header</kbd></dt>
<dd><p>Number of header rows to ignore. Default is 0</p>
<p>Default: 0</p>
</dd>
</dl>
</section>
</section>
<section id="interval-mds">
<h3>interval-mds<a class="headerlink" href="#interval-mds" title="Link to this heading">#</a></h3>
<p>Reads k-mer frequencies from a file and calculates a motif diversity score (MDS) for each interval using normalized Shannon entropy as described by Jiang et al (2020).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">interval</span><span class="o">-</span><span class="n">mds</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEP</span><span class="p">]</span> <span class="p">[</span><span class="n">file_path</span><span class="p">]</span> <span class="n">file_out</span>
</pre></div>
</div>
<section id="positional-arguments_repeat11">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat11" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>file_path</kbd></dt>
<dd><p>Tab-delimited or similar file containing one column for all k-mers a one column for frequency. Reads from stdin by default.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>file_out</kbd></dt>
<dd><p>Path to the output BED/BEDGraph file containing MDS for each interval.</p>
<p>Default: “-”</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat12">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat12" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-s, --sep</kbd></dt>
<dd><p>Separator used in tabular file.</p>
<p>Default: “      “</p>
</dd>
</dl>
</section>
</section>
<section id="filter-bam">
<h3>filter-bam<a class="headerlink" href="#filter-bam" title="Link to this heading">#</a></h3>
<p>Filters a BAM file so that all reads are in mapped pairs, exceed a certain MAPQ, are not flagged for quality, are read1, are not secondary or supplementary alignments, and are on the same reference sequence as the mate.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="nb">filter</span><span class="o">-</span><span class="n">bam</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">r</span> <span class="n">REGION_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUALITY_THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">hi</span> <span class="n">FRACTION_HIGH</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">-</span><span class="n">lo</span> <span class="n">FRACTION_LOW</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WORKERS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                         <span class="n">input_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat12">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat12" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><p>Path to BAM file.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat13">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat13" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-r, --region-file</kbd></dt>
<dd><p>Only output alignments overlapping the intervals in this BED file will be included.</p>
</dd>
<dt><kbd>-o, --output-file</kbd></dt>
<dd><p>Output BAM file path.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-q, --quality_threshold</kbd></dt>
<dd><p>Minimum mapping quality threshold.</p>
<p>Default: 30</p>
</dd>
<dt><kbd>-hi, --fraction-high</kbd></dt>
<dd><p>Maximum length for a fragment to be included in output BAM.</p>
</dd>
<dt><kbd>-lo, --fraction-low</kbd></dt>
<dd><p>Minimum length for a fragment to be included in output BAM.</p>
</dd>
<dt><kbd>-w, --workers</kbd></dt>
<dd><p>Number of worker processes.</p>
<p>Default: 1</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
</dd>
</dl>
</section>
</section>
<section id="agg-bw">
<h3>agg-bw<a class="headerlink" href="#agg-bw" title="Link to this heading">#</a></h3>
<p>Aggregates a bigWig signal over constant-length intervals defined in a BED file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">agg</span><span class="o">-</span><span class="n">bw</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span> <span class="n">MEDIAN_WINDOW_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="n">input_file</span> <span class="n">interval_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat13">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat13" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>input_file</kbd></dt>
<dd><blockquote>
<div><p>A bigWig file containing signals over the intervals specified in interval file.</p>
</div></blockquote>
</dd>
<dt><kbd>interval_file</kbd></dt>
<dd><p>Path to a BED file containing intervals over which signals were calculated over.</p>
</dd>
</dl>
</section>
<section id="named-arguments_repeat14">
<h4>Named Arguments<a class="headerlink" href="#named-arguments_repeat14" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>-o, --output-file</kbd></dt>
<dd><p>A wiggle file containing the aggregate signal over the intervals specified in interval file.</p>
<p>Default: “-”</p>
</dd>
<dt><kbd>-m, --median-window-size</kbd></dt>
<dd><p>Size of the median filter window used to adjust WPS scores. Only modify if aggregating WPS signals.</p>
<p>Default: 0</p>
</dd>
<dt><kbd>-v, --verbose</kbd></dt>
<dd><p>Enable verbose mode to display detailed processing information.</p>
</dd>
</dl>
</section>
</section>
<section id="gap-bed">
<h3>gap-bed<a class="headerlink" href="#gap-bed" title="Link to this heading">#</a></h3>
<p>Creates a BED4 file containing centromeres, telomeres, and short-arm intervals, similar to the gaps annotation track for hg19 found on the UCSC Genome Browser (Kent et al 2002). Currently only supports hg19, b37, human_g1k_v37, hg38, and GRCh38</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">finaletoolkit</span> <span class="n">gap</span><span class="o">-</span><span class="n">bed</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">{</span><span class="n">hg19</span><span class="p">,</span><span class="n">b37</span><span class="p">,</span><span class="n">human_g1k_v37</span><span class="p">,</span><span class="n">hg38</span><span class="p">,</span><span class="n">GRCh38</span><span class="p">}</span> <span class="n">output_file</span>
</pre></div>
</div>
<section id="positional-arguments_repeat14">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments_repeat14" title="Link to this heading">#</a></h4>
<dl class="option-list">
<dt><kbd>reference_genome</kbd></dt>
<dd><p>Possible choices: hg19, b37, human_g1k_v37, hg38, GRCh38</p>
<p>Reference genome to provide gaps for.</p>
</dd>
<dt><kbd>output_file</kbd></dt>
<dd><p>Path to write BED file to. If “-” used, writes to stdout.</p>
</dd>
</dl>
</section>
<p>Gap is used liberally in this command, and in the case hg38/GRCh38, may refer to regions where there no longer are gaps in the reference sequence.</p>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../user_guide/help.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Help</p>
      </div>
    </a>
    <a class="right-next"
       href="../api_reference/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">API</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments">Named Arguments</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Sub-commands">Sub-commands</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coverage">coverage</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat1">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#frag-length-bins">frag-length-bins</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat1">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat2">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#frag-length-intervals">frag-length-intervals</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat2">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat3">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cleavage-profile">cleavage-profile</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat3">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat4">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#wps">wps</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat4">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat5">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#adjust-wps">adjust-wps</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat5">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat6">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#delfi">delfi</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat6">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat7">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#delfi-gc-correct">delfi-gc-correct</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat7">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat8">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#end-motifs">end-motifs</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat8">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat9">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#interval-end-motifs">interval-end-motifs</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat9">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat10">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mds">mds</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat10">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat11">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#interval-mds">interval-mds</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat11">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat12">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#filter-bam">filter-bam</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat12">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat13">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#agg-bw">agg-bw</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat13">Positional Arguments</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat14">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gap-bed">gap-bed</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments_repeat14">Positional Arguments</a></li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">

  <div class="tocsection sourcelink">
    <a href="../../_sources/documentation/cli_reference/index.rst.txt">
      <i class="fa-solid fa-file-lines"></i> Show Source
    </a>
  </div>
</div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2024, EpiFluidLab.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.4.7.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.4.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>